File: 3623-count-number-of-trapezoids-i/3623-count-number-of-trapezoids-i.cpp
class Solution {
public:
    int countTrapezoids(vector<vector<int>>& points) {
        const long long MOD = 1000000007;
        unordered_map<int,long long> mp;
        mp.reserve(points.size()*2);
        for (auto &p : points) mp[p[1]]++;
        vector<long long> seg;
        seg.reserve(mp.size());
        for (auto &e : mp) {
            long long k = e.second;
            if (k >= 2) seg.push_back(k*(k-1)/2 % MOD);
        }
        long long sum = 0, ans = 0;
        for (long long v : seg) {
            ans = (ans + v * sum) % MOD;
            sum = (sum + v) % MOD;
        }
        return ans;
    }
};

File: 3623-count-number-of-trapezoids-i/3623-count-number-of-trapezoids-i.py
class Solution:
    def countTrapezoids(self, points: List[List[int]]) -> int:
        MOD = 10**9 + 7
        mp = {}
        for x, y in points:
            mp[y] = mp.get(y, 0) + 1
        seg = []
        for k in mp.values():
            if k >= 2:
                seg.append((k * (k - 1) // 2) % MOD)
        s = 0
        ans = 0
        for v in seg:
            ans = (ans + v * s) % MOD
            s = (s + v) % MOD
        return ans